text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,776 @@
+In an interactive CLI it makes sense to prompt the user for their input if these situations are detected, even if the user originally had `--no-rollback` specified.
+
+Now think of a different interactive mode: An app with a UI.
+All the same applies as before, but the implementation of the interaction will be different.
+And finally we also need to handle this scenario in non interactive environments.
+Although this is the easiest because we simply respect the value of the flag.
+
+`IoHost` formalizes how we deal with interaction.
+In our code base, we codify where we have decision points that potentially warrant a (different) response from the integrator and in extension the user.
+
+<img alt="Timeline of the communication between Toolkit and Integrator: First the Toolkit sends three messages. Then it sends a Request to which the integrator resonds. Finally two more messages are send from the Toolkit." src="../images/0300/iohost.png" height="300">
+
+We have an other requirement, that is in a similar space: Full control over all output.
+Think different interaction modes (none vs console vs app), but also rewriting the message text messages and re-prioritization w.r.t to log levels.
+
+Both together make up the interface of `IoHost`.
+See the readme above for more details.
+
+#### Messages and Requests
+
+The proposed `IoHost` design requires us to commit to a stable set of messages and requests send between [Programmatic Toolkit] and integrator.
+A detailed definition of what would constitute a breaking change is included in the readme above.
+
+To support integrations, we will need to publish a list of all messages.
+We will use static code analysis to automatically compile and publish this list without adding additional maintenance burden on to the team.
+
+#### Synth-time Errors
+
+This includes making changes to all errors thrown from the construct library.
+If an integrator includes the CDK App in the same process, errors can either be handled directly or will be rethrown.
+Cloud Executables that are going through a child process will eventually print all errors to stderr.
+For these we are going to detect errors from the output streams and convert them to JavaScript errors.
+The Toolkit will then throw them as before.
+
+#### Future Extensions
+
+This proposal focuses on a feature set required for GA.
+However the design allows for easy future extensions:
+
+- Additional actions can be migrated to the [Programmatic Toolkit]
+- The `IoHost` can be extended to include other I/O tasks like filesystem interactions.
+  This would allow implementations that are truly, in-memory like in browser environments.
+- The Toolkit setup can allow additional configuration, like providing a custom SDK or authentication configuration.
+
+### Is this a breaking change?
+
+No.
+
+### What alternative solutions did you consider?
```

Comment:
What about the option of not providing an event API at all, and instead just focus on the objects the CLI returns, and make sure they contain all the information necessary for customers to take actions?

That is:

```ts
// no hooks available on the `cdk` instance
const deployment = cdk.deploy(...);

// take the output and do something
const stacks = deployment.listStacks();

....
```

I'm not really advocating for this approach, but I want to understand exactly what use-cases this solution doesn't address.

Reply:
added an alternative